File: server/v1.flask/views/forum/post_search.py
from flask import Blueprint, render_template, jsonify, request
import scripts.utils.decorators
import scripts.utils.responses
import scripts.utils.structs
import scripts.utils.hashFunctions
import secrets
import datetime
import inspect
import scripts.utils.forumFuncs
import controllers.database
from dotenv import dotenv_values
config =  dotenv_values('../.env')


post_search = Blueprint('post_search', __name__)

@post_search.route(f'/{scripts.utils.structs.forumPath}/post/<int:id>', methods=['POST', 'OPTIONS'])   # DEL, GET was,'t working because axios wouldn't set content length
@scripts.utils.decorators.errorHandle
@scripts.utils.decorators.token_required
@scripts.utils.decorators.rateLimit
def forumPostViewHome(authPayload, *args, **kwargs):
  if request.method == 'POST':
    id = int(kwargs.get('id'))
    data = request.get_json()
    user_id = authPayload.get('payload').get('userId')
    username = authPayload.get('payload').get('username')
    role = authPayload.get('payload').get('role')
    if 'session_id' not in data:
      return scripts.utils.responses.build_bad_req()
    session_id = data.get('session_id')
    scripts.utils.forumFuncs.trackForumFunctionsCalled(username, session_id, inspect.stack()[0][3])
    
    pullForumQuery = '''
      SELECT 
        forum_posts.id,
        forum_posts.date,
        forum_users.forum_username,
        forum_posts.title,
        forum_posts.body,
        forum_user_id,
        forum_post_category.name
      from forum_posts
      INNER JOIN forum_post_category
        on forum_post_category.id = forum_posts.category_id
      INNER JOIN forum_users
        on forum_users.id = forum_posts.forum_user_id
      where
        forum_posts.id = ? and
        (forum_posts.visible = 1 or forum_posts.forum_user_id = ? or ? = "True");'''
    postRaw = controllers.database.conn.fetch(pullForumQuery, (id, user_id, (role==999)))

    if len(postRaw) != 1:
      return scripts.utils.responses.build_not_found()

    post = {
      'id' : postRaw[0][0],
      'date' : postRaw[0][1],
      'author': postRaw[0][2],
      'title': postRaw[0][3],
      'body': postRaw[0][4],
      'author_id': postRaw[0][5],
      'category': postRaw[0][6],
    }

    # if visible != 1 or role == 999:
    #   scripts.utils.responses.build_not_found()


    addForumViewQuery = 'INSERT INTO forum_post_views VALUES (?, ?, ?, ?);'
    controllers.database.conn.fetch(addForumViewQuery, (None, datetime.datetime.now(), user_id, id))

    # Distinct views??
    pullForumViewsQuery = 'SELECT count(*) from forum_post_views where ? = "None" and forum_post_id = ?;'
    postViewsRaw = controllers.database.conn.fetch(pullForumViewsQuery, ('None', id))[0][0]

    post = {
      **post,
      'views': postViewsRaw,
    }

    kwargs = {
      'success': True,
      'data': post,
    }
    res = jsonify(kwargs)
    return scripts.utils.responses.build_actual_response(res)
  elif request.method == 'OPTIONS': 
    return scripts.utils.responses.build_preflight_response()
  else:
    raise RuntimeError('Method not allowed')

File: server/v2.golang/router/routes/forum/post_search.go
package forum

import (
	"fmt"
	types "kooperlingohr/portfolio/Types"
	"kooperlingohr/portfolio/controllers/database"
	"kooperlingohr/portfolio/lib"
	"kooperlingohr/portfolio/router/middleware/responses"
	"kooperlingohr/portfolio/utils"
	"net/http"
	"strconv"
	"strings"
	"time"
)

func PostSearch(w http.ResponseWriter, r *http.Request) {
	if r.Method == http.MethodPost {
		urlComponents := strings.Split(r.URL.Path, "/")
		matches := urlComponents[(len(urlComponents))-1]
		fmt.Println("test")
		// re := regexp.MustCompile(`^.*\/([0-9]+)$`)
		// matches := re.FindStringSubmatch(r.URL.Path)
		// fmt.Println(matches)
		fmt.Println("test")

		// fmt.Println(routeVar)
		if len(matches) < 1 {
			// fail
		}

		postId := utils.HandleErrorDeconstruct(strconv.ParseInt(matches, 10, 16))

		decodedToken := r.Context().Value("decodedToken").(types.JWTbody)
		var body types.SessionId
		utils.ParseReqBody(r, &body)

		lib.TrackForumFunctionsCalled(decodedToken.Username, body.SessionID, fmt.Sprintf("/post/%d", postId))

		pullForumQuery := `
		SELECT 
			forum_posts.id as id,
			forum_posts.date as date,
			forum_users.forum_username as author,
			forum_posts.title as title,
			forum_posts.body as body,
			forum_posts.forum_user_id as author_id,
			forum_post_category.name as category
		from forum_posts
		INNER JOIN forum_post_category
			on forum_post_category.id = forum_posts.category_id
		INNER JOIN forum_users
			on forum_users.id = forum_posts.forum_user_id
		where
			forum_posts.id = ? and
			(forum_posts.visible = 1 or forum_posts.forum_user_id = ? or ? = true);`
		postRaw := utils.HandleErrorDeconstruct(database.ExecuteSQLiteQuery[types.ForumPostVerbose](pullForumQuery, []interface{}{postId, decodedToken.UserID, (decodedToken.Role == 999)}))

		if len(postRaw) != 1 {
			// fail special, 206
		}

		pullForumViewsQuery := "SELECT count(*) from forum_post_views where forum_post_id = ?;"
		pullForumViews := database.SimpleQuery[int64](pullForumViewsQuery, []interface{}{postId})

		resp := types.ForumPostResponseVerbose{
			Views: pullForumViews + 1,
			ForumPostVerbose: types.ForumPostVerbose{
				ID:       postRaw[0].ID,
				Date:     postRaw[0].Date,
				Author:   postRaw[0].Author,
				Title:    postRaw[0].Title,
				Body:     postRaw[0].Body,
				AuthorID: postRaw[0].AuthorID,
				Category: postRaw[0].Category,
			},
		}
		responses.BuildSuccessResponse(w, resp)

		now := time.Now()
		dt := now.Format(utils.GetTimeFormat())

		addForumViewQuery := `INSERT INTO forum_post_views VALUES (?, ?, ?, ?);`
		database.Insert(addForumViewQuery, []interface{}{nil, dt, decodedToken.UserID, postId})

		return
	} else if r.Method == http.MethodGet {
	}
}
